internal/dockerauth/api.go

```diff
@@ -107,23 +111,133 @@ type tokenResponse struct {
 }
 
-func (h *Handler) Token(p httprequest.Params, req *tokenRequest) (*tokenResponse, error) {
-	ras, err := ParseScope(req.Scope)
+func (h *Handler) handler(p httprequest.Params) (*handler, context.Context, error) {
+	store, err := h.params.Pool.RequestStore()
+	if err != nil {
+		return nil, p.Context, errgo.Mask(err)
+	}
+	return &handler{h, store}, p.Context, nil
+}
+
+type handler struct {
+	h     *Handler
+	store *charmstore.Store
+}
+
+func (h *handler) Close() error {
+	h.store.Close()
+	return nil
+}
+
+func (h *handler) Token(p httprequest.Params, req *tokenRequest) (*tokenResponse, error) {
+	ras, err := parseScope(req.Scope)
 	if err != nil {
 		logger.Infof("Invalid scope: %s", err)
 	}
-	var issuer string
-	if len(h.Certs) > 0 {
-		issuer = h.Certs[0].Subject.CommonName
+	ms := credentials(p.Request)
+	declared := checkers.InferDeclared(ms)
+	var user user = noUser{}
+	if identity, err := h.h.params.IDMClient.DeclaredIdentity(declared); err == nil {
+		if u, ok := identity.(*idmclient.User); ok {
+			user = u
+		}
+	}
+	filteredRAs := make([]resourceAccessRights, 0, len(ras))
+	for _, ra := range ras {
+		if ra.Type != "repository" {
+			continue
+		}
+		acl := h.repositoryACL(ra.Name)
+		filteredActions := make([]string, 0, len(ra.Actions))
+		for _, a := range ra.Actions {
+			switch a {
+			case "pull":
+				allow, _ := user.Allow(acl.Read)
+				if !allow {
+					continue
+				}
+			case "push":
+				allow, _ := user.Allow(acl.Write)
+				if !allow {
+					continue
+				}
+			default:
+				continue
+			}
+			if h.store.Bakery.Check(ms, checkers.New(declared)) != nil {
+				continue
+			}
+			filteredActions = append(filteredActions, a)
+		}
+		if len(filteredActions) == 0 {
+			continue
+		}
+		filteredRAs = append(filteredRAs, resourceAccessRights{
+			Type:    ra.Type,
+			Name:    ra.Name,
+			Actions: filteredActions,
+		})
 	}
 	issuedAt := time.Now()
-	expiresAt := issuedAt.Add(h.TokenValid)
+	s, err := h.createToken(filteredRAs, req.Service, issuedAt)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	return &tokenResponse{
+		Token:     s,
+		ExpiresIn: int(h.h.TokenValid / time.Second),
+		IssuedAt:  issuedAt,
+	}, nil
+}
+
+func credentials(req *http.Request) macaroon.Slice {
+	_, pw, _ := req.BasicAuth()
+	b, err := base64.RawStdEncoding.DecodeString(pw)
+	if err != nil {
+		logger.Debugf("invalid macaroon: %s", err)
+		return nil
+	}
+	var ms macaroon.Slice
+	if err := ms.UnmarshalBinary(b); err != nil {
+		logger.Debugf("invalid macaroon: %s", err)
+		return nil
+	}
+	return ms
+}
+
+func (h *handler) repositoryACL(name string) mongodoc.ACL {
+	parts := strings.SplitN(name, "/", 4)
+	if len(parts) != 4 {
+		return mongodoc.ACL{}
+	}
+	baseURL := charm.URL{
+		Schema:   "cs",
+		User:     parts[0],
+		Name:     parts[1],
+		Revision: -1,
+	}
+	be, err := h.store.FindBaseEntity(&baseURL, map[string]int{
+		"_id":         1,
+		"channelacls": 1,
+	})
+	if err != nil {
```

Comment:
Again, I don't like the fact that we're just ignoring errors. If there's a mongo error, we should return it, not fail silently.